File: week6/pset6/sentimental-credit/credit.py
# TODO

from cs50 import get_int


def main():

    # Get user input
    number = get_int("Number: ")

    # Determine the length of the card number
    length = 1
    while True:
        i = number / pow(10, length)
        if i >= 1:
            length += 1
        else:
            break

# Checks length and get_digits work
    # print(f"Length: {length}")
    # which = get_int("What digit to last do you want to see? ")
    # print(f"The {which}th to last digit is: {get_digits(number, which)}")

    # Checksum
    check_sum = 0
    # First half
    for j in range(2, 17, 2):
        check_sum += get_digits((get_digits(number, j) * 2), 1) + get_digits((get_digits(number, j) * 2), 2)
    # Second half
    for k in range(1, 16, 2):
        check_sum += get_digits(number, k)

    # Validate checksum
    if (check_sum % 10 == 0):
        # AMEX check
        if length == 15 and get_digits(number, length) == 3 and (get_digits(number, length - 1) == 4
                                                                 or get_digits(number, length - 1) == 7):
            print("AMEX")
        # MASTERCARD check
        elif length == 16 and get_digits(number, length) == 5 and (get_digits(number, length - 1) == 1
                                                                   or get_digits(number, length - 1) == 2
                                                                   or get_digits(number, length - 1) == 3
                                                                   or get_digits(number, length - 1) == 4
                                                                   or get_digits(number, length - 1) == 5):
            print("MASTERCARD")
        # VISA check
        elif (length == 13 or length == 16) and get_digits(number, length) == 4:
            print("VISA")
        # Else
        else:
            print("INVALID")
    # If checksum is bad
    else:
        print("INVALID")


# Asks for number and position of the digit (to last), returns the digit
def get_digits(number, position):
    digit = number % pow(10, position) // pow(10, position - 1)
    return digit


main()

File: week1/pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

// Returns true if the card number is valid
bool validate(long long card_number)
{
    int sum = 0;
    int num_digits = floor(log10(card_number)) + 1;
    int parity = num_digits & 1;
    for (int i = 0; i < num_digits; i++)
    {
        int digit = card_number % 10;
        if (i % 2 == parity)
        {
            digit *= 2;
            if (digit > 9)
            {
                digit -= 9;
            }
        }
        sum += digit;
        card_number /= 10;
    }
    return sum % 10 == 0;
}

// Returns the type of the card as a string
const char *type(long long card_number)
{
    if ((card_number >= 34e13 && card_number < 35e13) ||
        (card_number >= 37e13 && card_number < 38e13))
    {
        return "AMEX\n";
    }
    else if (card_number >= 51e14 && card_number < 56e14)
    {
        return "MASTERCARD\n";
    }
    else if ((card_number >= 4e12 && card_number < 5e12) ||
             (card_number >= 4e15 && card_number < 5e15))
    {
        return "VISA\n";
    }
    else
    {
        return "INVALID\n";
    }
}

int main(void)
{
    long long card_number = get_long_long("Number: ");
    if (validate(card_number))
    {
        printf("%s", type(card_number));
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}